v2/googlecloud-to-googlecloud/src/test/java/com/google/cloud/teleport/v2/templates/spannerchangestreamstobigquery/SpannerChangeStreamsToBigQueryIT.java

```diff
@@ -0,0 +1,224 @@
+
+  private LaunchInfo launchInfo;
+
+  @Before
+  public void setup() throws IOException {
+    spannerResourceManager =
+        SpannerResourceManager.builder(testName, PROJECT, REGION).maybeUseStaticInstance().build();
+    bigQueryResourceManager =
+        BigQueryResourceManager.builder(testName, PROJECT, credentials).build();
+  }
+
+  @After
+  public void tearDown() {
+    ResourceManagerUtils.cleanResources(spannerResourceManager, bigQueryResourceManager);
+  }
+
+  @Test
+  public void testSpannerChangeStreamsToBigQueryBasic() throws IOException {
+    String spannerTable = testName + RandomStringUtils.randomAlphanumeric(1, 5);
+    String createTableStatement =
+        String.format(
+            "CREATE TABLE %s (\n"
+                + "  Id INT64 NOT NULL,\n"
+                + "  FirstName String(1024),\n"
+                + "  LastName String(1024),\n"
+                + ") PRIMARY KEY(Id)",
+            spannerTable);
+
+    String cdcTable = spannerTable + "_changelog";
+    spannerResourceManager.executeDdlStatement(createTableStatement);
+    String createChangeStreamStatement =
+        String.format(
+            "CREATE CHANGE STREAM %s_stream FOR %s OPTIONS (value_capture_type = 'NEW_ROW')",
+            testName, spannerTable);
+    spannerResourceManager.executeDdlStatement(createChangeStreamStatement);
+    bigQueryResourceManager.createDataset(REGION);
+
+    Function<LaunchConfig.Builder, LaunchConfig.Builder> paramsAdder = Function.identity();
+    ;
+
+    launchInfo =
+        launchTemplate(
+            paramsAdder.apply(
+                LaunchConfig.builder(testName, specPath)
+                    .addParameter("spannerProjectId", PROJECT)
+                    .addParameter("spannerInstanceId", spannerResourceManager.getInstanceId())
+                    .addParameter("spannerDatabase", spannerResourceManager.getDatabaseId())
+                    .addParameter(
+                        "spannerMetadataInstanceId", spannerResourceManager.getInstanceId())
+                    .addParameter("spannerMetadataDatabase", spannerResourceManager.getDatabaseId())
+                    .addParameter("spannerChangeStreamName", testName + "_stream")
+                    .addParameter("bigQueryDataset", bigQueryResourceManager.getDatasetId())
+                    .addParameter("rpcPriority", "HIGH")
+                    .addParameter("disableDlqRetries", "true")));
+
+    assertThatPipeline(launchInfo).isRunning();
+
+    int key = nextValue();
+    String firstName = UUID.randomUUID().toString();
+    String lastName = UUID.randomUUID().toString();
+    List<Pair> colValPairs =
+        Arrays.asList(new Pair("FirstName", firstName), new Pair("LastName", lastName));
+    Mutation expectedData = generateTableRow(spannerTable, key, colValPairs);
+    spannerResourceManager.write(Collections.singletonList(expectedData));
+    String query = queryCdcTable(cdcTable, key);
+    waitForQueryToReturnRows(query, 1, true);
```

Comment:
We already have helper test classes to perform this function, in this case, `BigQueryRowsCheck`
```suggestion
    Result result =
        pipelineOperator()
            .waitForConditionsAndFinish(
                createConfig(info),
                BigQueryRowsCheck.builder(bigQueryResourceManager, cdcTable)
                    .setMinRows(1)
                    .build());

    assertThatResult(result).meetsConditions();
```

The only difference is the query you were using specified an Id equality clause, but that doesn't seem relevant as there should only be one row in the table anyway since it us spun up as a fresh instance, but correct me if I'm wrong

Reply:
Thanks for sharing this test util function! Yes you are correct that for this base case test, it doesn't really matter because only one row will be written to changelog table. However this pr is only the first step for more complex tests in later prs. I should have provided more context on the follow up prs, sorry for the confusions.

 For example, this is the draft of the follow up test pr(in a separate forked repo): https://github.com/ChangyuLi28/DataflowTemplates/commit/a1c2872a700a410fed7c95193a0c4103a3fa55df#diff-28b3790277fb958cc3acef6775b2e789ad279e97d8f962068915c9800a647869 
For example, we will have new test case that tests the schema update handling during pipeline running with the above process:
1. Start the pipeline
2. Insert a new row with id = 1
3. Verify table row in BigQuery with id=1 is correct
4. Add a new column to the spanner table
5. Insert a new row with id = 2 and the newly added column specified
6. Verify table row in BigQuery with id=2 is correct and the newly added column is correctly populated.

In this scenario, we will rely on this `queryCdcTable` function to only query the row with id=2, which is the row after the schema update. Thus I thought it would be more consistent for all tests in this file to use this same test util function to query the changelog table.

Reply:
I did something very similar in the Datastream IT's (also does CDC)

Take for example https://github.com/GoogleCloudPlatform/DataflowTemplates/blob/main/v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/DataStreamToSpannerIT.java

In particular https://github.com/GoogleCloudPlatform/DataflowTemplates/blob/0fef11cae7308a6389bda6cafb6fc7c9b68987ce/v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/DataStreamToSpannerIT.java#L406-L424

For your use case, you could do something like
```
ChainedConditionCheck conditionCheck = 
     ChainedConditionCheck.builder( 
             List.of( 
                 writeSpannerData(...), 
                 BigQueryRowsCheck.builder(bigQueryResourceManager, cdcTable)
                    .setMinRows(1)
                    .build(), 
                 writeSpannerData(...), 
                 BigQueryRowsCheck.builder(bigQueryResourceManager, cdcTable)
                    .setMinRows(2)
                    .build())) 
         .build(); 
```

where `writeSpannerData` is a `ConditionCheck` so that it can be chained in a particular order with the actual checks.

Ultimately, as long as the test doesn't break PR submissions or add burden to the core template infra, it is up to you and your team how the IT's are implemented. I just wanted to lay out some patterns that we like to use to possibly make maintenance a bit easier

Reply:
Thanks for the code pointer! I definitely agree that the chained conditional check is a more elegant and clean way for this scenario. However there is a key difference between datastream to spanner it and spanner cdc to big query it. 

`BigQueryRowsCheck.builder` takes in a TableId as parameter.
 For our case, the corresponding cdc table in big query sink is implicitly created by the template:
![image](https://github.com/GoogleCloudPlatform/DataflowTemplates/assets/43364656/8352deaf-16fc-4327-a3f7-a1c3b7cf41ff)
If we directly get the table id like this for the cdcTable, it failed with table not exist error, since it could take couple minutes to start the pipeline and internally create the cdc table(I verified this locally). Thus we will not only need to wait until the row is populated in the cdc table, but also need to wait for the table itself is created behind the scene. That's why I caught the table not found exception in the `dataShownUp` function:
![image](https://github.com/GoogleCloudPlatform/DataflowTemplates/assets/43364656/fa22a895-e6a8-4566-be93-ee730d6fb946)
Where as in DatastreamToSpannerIT, the sink tables in Spanner are [manually created at first](https://github.com/GoogleCloudPlatform/DataflowTemplates/blob/0fef11cae7308a6389bda6cafb6fc7c9b68987ce/v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/DataStreamToSpannerIT.java#L367) and the spanner row checker only takes a table name string parameter.

Reply:
You also don't have to call it that way. 

Instead of 
```
TableId cdcTableId = bigQueryResourceManager.getTableIfExists(cdcTable).getTableId();
```

you could just use
```
TableId cdcTableId = TableId.of(bigQueryResourceManager.getDatasetId(), cdcTable);
```

The `BigQueryRowsCheck` catches exceptions and throws a WARN instead, so as it tries to query the table size, it will throw a warning that the table does not exist each cycle. We have lots of examples of templates that implicitly create BQ tables (especially DLQ tables), and they use the check like that.